common/core/web/keyboard-processor/src/text/codes.ts

```diff
@@ -9,6 +9,7 @@ namespace com.keyman.text {
       "SHIFT":0x0010,
       "CTRL":0x0020,
       "ALT":0x0040,
+      "META":0x0080, // Represents command keys, which some OSes use for shortcuts we don't want to block.
```

Comment:
I don't think this is the appropriate place to define. We do not intend to ever allow keyboards to override the meta (macOS Command) key, right? These values are defined by the compiler and intended for use by Keyman keyboards.

Furthermore, these modifier bit flags are defined in a number of places, including Keyman Developer, Keyman Core (Linux), Keyman for Windows and Keyman for Mac. I am very hesitant to add a new value without verifying throughout what the impact is.

Note that in the Windows world, this bitfield is used as a 16 bit WORD smushed into a DWORD together with a (WORD) VKEY, so we cannot use higher values either.

Reply:
So... we should handle this key differently from every other key on the keyboard?

> We do not intend to ever allow keyboards to override the meta (macOS Command) key, right?

I mean... is that right?  The only reason that may hold true is that the other major OSes don't give us access to the key, right?

Despite that, my main counterargument:  https://github.com/keymanapp/keyman/blob/297efbbaf0e952f4d88f9b24b45d2d60e325e61e/common/core/web/keyboard-processor/src/text/codes.ts#L4-L19

For crying out loud, we model _`SCROLL_LOCK`_ / _`NO_SCROLL_LOCK`_.  I mean, I guess we _theoretically_ allow keyboards to use that.  Though... [`SCROLL_LOCK` isn't even supported in Web on Mac](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState#Modifier_keys_on_Gecko).  (Well... at least within Firefox.  I can't find anything clear for Chrome or Safari.)

In the list of keycodes, we even bother tracking some pretty rarely-used keys:  https://github.com/keymanapp/keyman/blob/297efbbaf0e952f4d88f9b24b45d2d60e325e61e/common/core/web/keyboard-processor/src/text/codes.ts#L42

I couldn't even tell you what `K_SEL` and `K_EXEC` represent... and we still model _them_.

----

I'm not asking us to make this a full, cross-platform thing, at least not right now.  To be extra-clear, I'm **absolutely** against adding any Meta-based keyboard rules for 14.0.  It's just that this is the cleanest _and most internally-consistent_ solution.  I could always modify the comment to note that it's a "tentative" value used only within Web for now, as well.

>  Note that in the Windows world, this bitfield is used as a 16 bit WORD smushed into a DWORD together with a (WORD) VKEY, so we cannot use higher values either.

Are you saying that this would fill in the last available modifier bit-flag slot?  That nothing over `0x4000` (`VIRTUAL_KEY`) is usable?  Even if it's actually "nothing over `0x8000`... is there some other modifier set we expect to come into being?  I'm assuming this (the limited extra bit-space) is your concern, as otherwise I don't see how this part is relevant - `0x0080` is safely tucked well within the supported bitfield.

Reply:
> So... we should handle this key differently from every other key on the keyboard?
> 
> > We do not intend to ever allow keyboards to override the meta (macOS Command) key, right?
> 
> I mean... is that right? The only reason that may hold true is that the other major OSes don't give us access to the key, right?

The corresponding key on Windows is the <kbd>Windows</kbd> key. We don't provide access to that either. Both the <kbd>Command</kbd> key and the <kbd>Windows</kbd> key are never intended to be used to provide character output. Unlike the other modifiers, they have never been used for that purpose (although <kbd>Win</kbd>+<kbd>n</kbd> for example in Win10 2004 generates `n`, because there is no <kbd>Win</kbd>+<kbd>n</kbd> shortcut currently defined, there is no guarantee that <kbd>Win</kbd>+<kbd>n</kbd> won't be used in the future).

Regarding scroll lock, `K_SEL`, etc, there are a whole bunch of "interesting" legacy choices in the code but that's not necessarily a good argument for future changes. Over time, we've become slightly more prescriptive in our keyboard design philosophy -- for example, on desktop keyboards we recommend using only Shift, RAlt and Shift+RAlt modifiers for new keyboards (where [R]Alt corresponds to the [R]Option key on Mac), even though older keyboards did all manner of wonderful things with Ctrl etc. This reflects increased understanding of cross platform and cross application compatibility concerns (e.g. Ctrl+C is not a good choice for character output!).

Don't misunderstand me: I am not against encoding the meta/command key in the bitfield in the future. I just don't want us to do that without a proper cross-platform design phase. For example, are Left and Right Meta separate? Note that `0x8000` is defined as `VIRTUALCHARKEY` on Windows (and should be defined here as well so we don't treat it as 'available').

Similarly, if we made this change, `K_MODIFIERFLAG` would need to be updated to `0xFF` in the cpp headers (but don't do this for now).

I guess I'd be okay with this change, if in the source we do make a note about it being tentative (with reference to this discussion), and add a corresponding note in common\core\desktop\src\kmx\kmx_file.h, mac\KeymanEngine4Mac\KeymanEngine4Mac\KME\KMBinaryFileFormat.h and windows\src\global\inc\Compiler.h (compv4.h is not relevant, only describes Keyman 4 compiled kmx keyboards).

So long as it is documented and tracked in this way, it isn't difficult to roll it back if we decide on another solution for meta in the longer term.

Reply:
> (although Win+n for example in Win10 2004 generates `n`, because there is no Win+n shortcut currently defined, there is no guarantee that Win+n won't be used in the future).

Honestly, sounds like Keyman for Windows could use a similar solution to whatever we run with in the long-term.  Sounds like Keyman Engine (in general) should proactively avoid rule processing in `Meta` scenarios.

Also, a minor note from my documentation searches:  the Windows key is interpreted as `Meta` in some cases - some, as some browsers selectively ignore the Windows case, never reporting it.  Also, since we have no desire to allow rules targeting `Meta`, I see no point in `MetaLeft` vs `MetaRight`.  `Meta` itself should be good for bypassing rules.

> Note that 0x8000 is defined as VIRTUALCHARKEY on Windows (and should be defined here as well so we don't treat it as 'available').

:+1:  Always good to convert headspace documentation into actual documentation when possible.

------

I'm 90% leaning toward continuing the current strategy and documenting this approach per this discussion.  Admittedly, there is a different approach I'm still slightly considering.

That 10% "other approach": maybe there's another way to "abort" the keystroke safely within Web.  That pathway forward feels less clear to me at the moment though - there are a lot of event handlers driving the KMW engine, and thus assumptions on how those handlers may interact.

Reply:
> Honestly, sounds like Keyman for Windows could use a similar solution to whatever we run with in the long-term. Sounds like Keyman Engine (in general) should proactively avoid rule processing in `Meta` scenarios.

Keyman on Windows doesn't need to do anything here. The Win+n is a Windows thing, not a Keyman thing.

I will reiterate that as we don't plan to support the meta/command/windows key, the long term solution is to avoid processing it altogether rather than add it to the modifier bitmask.

However, as discussed, the short term solution we agreed on in the planning meeting is adequate, if a cleaner solution is too much work: we will use the meta bitflag here for now and remove it when we have the time to engineer a cleaner solution.

Keyman started on Windows so there are many Windowsisms that have leaked across the platform boundary. But I really want to try to avoid having platform-specific patches like this one leaking across those platform boundaries in the future.

> 👍 Always good to convert headspace documentation into actual documentation when possible.

That was of course already in the headers for other the platforms and I noted it when I compared the defined values, rather than it being 'headspace documentation'.